resolver/map.go

```diff
@@ -136,3 +136,134 @@ func (a *AddressMap) Values() []any {
 	}
 	return ret
 }
+
+type endpointNodeList []endpointNode
+
+type endpointNode struct {
+	addrs map[string]int
+	val   any
+}
+
+// Equal returns whether the unordered set of addrs counts are the same between
+// the endpoint nodes.
+func (en endpointNode) Equal(en2 endpointNode) bool {
+	if len(en.addrs) != len(en2.addrs) {
+		return false
+	}
+	for addr, count := range en.addrs {
+		if count2, ok := en2.addrs[addr]; !ok || count != count2 {
+			return false
+		}
+	}
+	return true
+}
+
+func toEndpointNode(endpoint Endpoint) endpointNode {
+	en := make(map[string]int)
+	for _, addr := range endpoint.Addresses {
+		if count, ok := en[addr.Addr]; ok {
+			en[addr.Addr] = count + 1
+		}
+		en[addr.Addr] = 1
+	}
+	return endpointNode{
+		addrs: en,
+	}
+}
+
+// EndpointMap is a map of endpoints to arbitrary values keyed on only the
+// unordered set of addresses string within an endpoint. The reason this uses a
+// slice instead of a map is because slices or maps cannot be used as map keys.
```

Comment:
External documentation should not include implementation details or rational for implementation details unless it's important to the user of the type.

Reply:
Ah, ok noted. Moved the comment to the field (which is unexported and users of this package won't see it).